hera_cal/redcal.py

```diff
@@ -269,6 +337,127 @@ def multiply_by_gains(target, gains, target_type='vis'):
+                are not to be solved for)
+
+        Returns:
+            None
+        """
+        linsolve.LinProductSolver.__init__(self, data, sol0, wgts=wgts, **kwargs)
+        self.gain = np.float32(gain) # float32 to avoid accidentally promoting data to doubles.
+
+    def _get_ans0(self, sol, keys=None):
+        '''Evaluate the system of equations given input sol. 
+        Specify keys to evaluate only a subset of the equations.'''
+        if keys is None:
+            keys = self.keys
+        _sol = {k+'_':v.conj() for k,v in sol.items() if k.startswith('g')}
+        _sol.update(sol)
+        return {k: eval(k, _sol) for k in keys}
+
+    def solve_iteratively(self, conv_crit=1e-10, maxiter=50, check_every=4, check_after=1, verbose=False):
+        """Repeatedly solves and updates solution until convergence or maxiter is reached. 
+        Returns a meta object containing the number of iterations, chisq, and convergence criterion.
+
+        Args:
+            conv_crit: A convergence criterion (default 1e-10) below which to stop iterating. 
+                Converegence is measured L2-norm of the change in the solution of all the variables
+                divided by the L2-norm of the solution itself.
+            maxiter: An integer maximum number of iterations to perform before quitting. Default 50.
+            check_every: Compute convergence every Nth iteration (saves computation).  Default 4.
+            check_after: Start computing convergence only after N iterations.  Default 1.
+
+        Returns: meta, sol
+            meta: a dictionary with metadata about the solution, including
+                iter: the number of iterations taken to reach convergence (or maxiter)
+                chisq: the chi^2 of the solution produced by the final iteration
+                conv_crit: the convergence criterion evaluated at the final iteration
+            sol: a dictionary of complex solutions with variables as keys
+        """
+        sol = self.sol0
+        terms = [(linsolve.get_name(gi),linsolve.get_name(gj),linsolve.get_name(uij)) 
+            for term in self.all_terms for (gi,gj,uij) in term]
+        dmdl_u = self._get_ans0(sol)
+        chisq = sum([np.abs(self.data[k]-dmdl_u[k])**2 * self.wgts[k] for k in self.keys])
+        # variables with '_u' are flattened and only include pixels that need updating
+        dmdl_u = {k:v.flatten() for k,v in dmdl_u.items()}
+        # wgts_u hold the wgts the user provides.  dwgts_u is what is actually used to wgt the data
+        wgts_u = {k: (v * np.ones(chisq.shape, dtype=np.float32)).flatten() for k,v in self.wgts.items()}
+        sol_u = {k:v.flatten() for k,v in sol.items()}
+        iters = np.zeros(chisq.shape, dtype=np.int)
+        conv = np.ones_like(chisq)
+        update = np.where(chisq > 0)
+        for i in range(1,maxiter+1):
+            if verbose: print('Beginning iteration %d/%d' % (i,maxiter))
+            if (i % check_every) == 1:
```

Comment:
Shouldn't this be `if (i == 1) or (i % check_every) == 1:`? Because otherwise, if `check_every` is 1, it'll never run.